packages/flutter/lib/src/widgets/widget_span.dart

```diff
@@ -0,0 +1,262 @@
+        alignment,
+        baseline: currentDimensions.baseline,
+        baselineOffset: currentDimensions.baselineOffset,
+      );
+    }
+    if (hasStyle)
+      builder.pop();
+  }
+
+  /// Calls visitor on this [WidgetSpan]. There are no children spans to walk.
+  bool visitChildren(bool visitor(InlineSpan span)) {
+    if (!visitor(this))
+      return false;
+    if (children != null) {
+      for (InlineSpan child in children) {
+        if (!child.visitChildren(visitor))
+          return false;
+      }
+    }
+    return true;
+  }
+
+  String toPlainText() {
+    return widget.toString();
+  }
+
+  int codeUnitAt(int index) {
+    return null;
+  }
+
+  /// Describe the difference between this widget span and another [TextSpan],
+  /// in terms of how much damage it will make to the rendering. The comparison
+  /// is deep.
+  ///
+  /// Comparing a [WidgetSpan] with a [TextSpan] will result in [RenderComparison.layout].
+  ///
+  /// See also:
+  ///
+  ///  * [TextStyle.compareTo], which does the same thing for [TextStyle]s.
+  RenderComparison compareTo(InlineSpan other) {
+    if (identical(this, other))
+      return RenderComparison.identical;
+    if (other.runtimeType is! WidgetSpan)
+      return RenderComparison.layout;
+    final WidgetSpan typedOther = other;
+    if (typedOther.widget != widget)
+      return RenderComparison.layout;
+    if ((style == null) != (other.style == null))
+      return RenderComparison.layout;
+    RenderComparison result = RenderComparison.identical;
+    if (style != null) {
+      final RenderComparison candidate = style.compareTo(other.style);
+      if (candidate.index > result.index)
+        result = candidate;
+      if (result == RenderComparison.layout)
+        return result;
+    }
+    // WidgetSpans are always a leaf node in the TextSpan tree, and have
+    // no children.
+    return result;
+  }
+
+  @override
+  bool operator ==(dynamic other) {
+    if (identical(this, other))
+      return true;
+    if (other.runtimeType != runtimeType)
+      return false;
+    final WidgetSpan typedOther = other;
+    return typedOther.widget == widget
+        && typedOther.style == style;
+  }
+
+  /// Returns the text span that contains the given position in the text.
+  TextSpan getSpanForPosition(TextPosition position) {
+    assert(debugAssertIsValid());
+    return null;
+  }
+
+
+  /// In checked mode, throws an exception if the object is not in a
+  /// valid configuration. Otherwise, returns true.
+  ///
+  /// This is intended to be used as follows:
+  ///
+  /// ```dart
+  /// assert(myTextSpan.debugAssertIsValid());
+  /// ```
+  bool debugAssertIsValid() {
+    assert(() {
+      if (!visitChildren((InlineSpan span) {
+        if (span is WidgetSpan)
+          return (span as WidgetSpan).widget != null;
+        TextSpan textSpan = span;
+        if (textSpan.children != null) {
+          for (InlineSpan child in textSpan.children) {
+            if (child == null)
```

Comment:
again this seems like something we could just verify in the TextSpan constructor. If we can't because it has to be const, then I would put this logic on a method on TextSpan itself. That would mean we don't have to have an `is` here either.

Reply:
wait, there _is_ an InlineSpan.debugAssertIsValid. Why do we need this at all?

Reply:
never mind, I see, this is the implementation on WidgetSpan.